Deduplicate Shard Started Requests

original-brownbear opened the pull request:
Deduplicate shard started requests the same way we deduplicate shard-failed
and shard snapshot state updates already.

closes #81628

original-brownbear committed:
Deduplicate Shard Started Requests

Deduplicate shard started requests the same way we deduplicate shard-failed
and shard snapshot state updates already.

closes #81628

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -81,7 +81,7 @@
     // a list of shards that failed during replication
     // we keep track of these shards in order to avoid sending duplicate failed shard requests for a single failing shard.
-    private final ResultDeduplicator<FailedShardEntry, Void> remoteFailedShardsDeduplicator = new ResultDeduplicator<>();
+    private final ResultDeduplicator<TransportRequest, Void> remoteFailedShardsDeduplicator = new ResultDeduplicator<>();
```
I think this field needs a rename now.

original-brownbear replied:
++ renamed and fixed comment

henningandersen reviewed (commented):
Can we add a test demonstrating that this works, both when master has stuff queued, causing the retries and when master fails over?

I wonder if we should change our logic here to always send to a new master to speed up recovery after a very slow/gc hung master is taken over by another master?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 81628

original-brownbear committed:
trivial test and dedup cleanup on failover

original-brownbear commented:
> Can we add a test demonstrating that this works, both when master has stuff queued, causing the retries and when master fails over?

I added a rather trivial test in the style of the tests that already exist for this thing (the test is good enough to demonstrate proper deduplication of requests IMO). Couldn't find a quick way of testing the thing below.

> I wonder if we should change our logic here to always send to a new master to speed up recovery after a very slow/gc hung master is taken over by another master?

Yea we had the same issue in shard snapshot state updates and I implemented the same solution now. Unfortunately I couldn't find a neat way of testing this quickly. In snapshotting this is a lot easier to test with the existing test infrastructure.
Not sure it's worth the effort to add a test for this today since it's certainly better than it was before by adding the clearing out of the dedup?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -196,15 +195,24 @@ public void remoteShardFailed(
+     * to the new master right away on master failover.
+     */
+    public void clearRemoteShardRequestDeduplicator() {
+        remoteShardStateUpdateDeduplicator.clear();
```
Since we call this from multiple threads, there is a bit of best-effort over this method, I think that is worth documenting.

For instance, this may clear out a remote shard failed request deduplication to the new master in edge cases. This does no real harm, since we still protect the master.

original-brownbear replied:
++ added

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java:
```diff
@@ -382,6 +382,33 @@ public void onFailure(Exception e) {
         assertThat(transport.capturedRequests(), arrayWithSize(0));
     }
 
+    public void testDeduplicateRemoteShardStarted() throws InterruptedException {
```
Can we either add a test or randomly clear the deduplicator here and then validate we see two requests at the end?

original-brownbear replied:
++ done

henningandersen reviewed (approved):
LGTM, with two small additions.

original-brownbear committed:
udpate comment + enhance test

original-brownbear commented:
Thanks Henning!

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/elasticsearch/pull/82106)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/82106">](https://github.com/elastic/elasticsearch/pull/82106) |

DaveCTurner commented:
LGTM as a small/interim fix, but fundamentally we should be using edge-triggering for the shard state transitions with appropriate failure handling to organise retries (see also #81626). Today’s level-triggered system was necessary when cluster state updates could be lost I guess, but that’s no longer the case. I opened https://github.com/elastic/elasticsearch/issues/82185 to track this tech debt.
